NitroxServer/ConsoleCommands/Abstract/Command.cs

```diff
@@ -1,62 +1,163 @@
 {
     public abstract class Command
     {
-        public string Name { get; protected set; }
-        public string[] Alias { get; protected set; }
-        public string Description { get; protected set; }
-        public string ArgsDescription { get; protected set; }
-        public Perms RequiredPermLevel { get; protected set; }
+        private int optional;
+        private int required;
+
+        public string Name { get; }
+        public string Description { get; }
+        public List<string> Alias { get; }
+        private string[] Args { get; set; }
+        public Perms RequiredPermLevel { get; }
+        public bool AllowedArgOverflow { get; }
+        public List<IParameter<object>> Parameters { get; }
 
-        protected Command(string name, Perms requiredPermLevel, string argsDescription, string description) : this(name, requiredPermLevel, argsDescription, "", null)
+        public Command(string name, Perms perm, string description, bool allowedArgOveflow = false)
         {
-            ArgsDescription = argsDescription;
+            Validate.NotNull(name);
+
             Name = name;
-            Description = description;
+            RequiredPermLevel = perm;
+            Alias = new List<string>();
+            Parameters = new List<IParameter<object>>();
+            AllowedArgOverflow = allowedArgOveflow;
+            Description = string.IsNullOrEmpty(description) ? "No description provided" : description;
         }
 
-        protected Command(string name, Perms requiredPermLevel, string argsDescription, string description, string[] alias)
+        protected abstract void Execute(Optional<Player> sender);
+
+        public void TryExecute(string[] args, Optional<Player> sender)
         {
-            Validate.NotNull(name);
-            Validate.NotNull(argsDescription);
-            Validate.NotNull(description);
+            if (args.Length < required)
+            {
+                SendMessage(sender, $"Error: Invalid Parameters\nUsage: {ToHelpText(true)}");
+                return;
+            }
 
-            Name = name;
-            Description = string.IsNullOrEmpty(description) ? "No description" : description;
-            ArgsDescription = argsDescription;
-            RequiredPermLevel = requiredPermLevel;
-            Alias = alias ?? new string[0];
+            if (!AllowedArgOverflow && args.Length > optional + required)
+            {
+                SendMessage(sender, $"Error: Too many Parameters\nUsage: {ToHelpText(true)}");
+                return;
+            }
+
+            try
+            {
+                Args = args;
+                Execute(sender);
+            }
+            catch (IllegalArgumentException e)
+            {
+                SendMessage(sender, $"Error: {e.Message}");
+            }
+            catch (Exception e)
+            {
+                Log.Error("Fatal error while trying to execute the command", e);
+            }
+
+            Args = null;
```

Comment:
Do we need to make this class stateful?  Can this be passed into Execute instead?

Reply:
Yes It used to be passed into Execute() but I decided in this PR to remove the ability to iterate on the original string[] produced by the Processor

Reply:
I'm working on refactoring this to use `CallArgs` subclass that will be the single argument passed to `Command.Execute` containing all the properties (`Sender` and `Args`) with `ReadAt(int)` methods.

Reply:
Done